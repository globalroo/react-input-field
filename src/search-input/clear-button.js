import * as React from "react";
import PropTypes from "prop-types";
import "font-awesome/css/font-awesome.min.css";
import { styles, BUTTON_WIDTH } from "src/search-input/styles/styles";

export const ClearButton = ({ visible, onClick, hovered }) => (
  <button
    onClick={onClick}
    style={{
      width: BUTTON_WIDTH,
      border: 0,
      verticalAlign: "middle",
      opacity: visible ? 1 : 0
    }}
  >
    <span
      style={hovered ? styles.clearButtonIconHover : styles.clearButtonIcon}
      className="clear-button-icon fa-stack fa-lg"
    >
      <i
        style={{ opacity: hovered ? 0 : 1 }}
        className="fa fa-circle-o fa-stack-2x"
      />
      <i className="fa fa-times fa-stack-1x" />
    </span>
  </button>
);

ClearButton.propTypes = {
  hovered: PropTypes.bool,
  onClick: PropTypes.func,
  visible: PropTypes.bool
};
