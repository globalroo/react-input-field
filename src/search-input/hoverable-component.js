import * as React from "react";
import PropTypes from "prop-types";

/*
** Adds mouse in / out state indicator to any child components
** contained within.
*/
export class Hoverable extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			hovered: false
		};
	}
	setRef = node => (this.ref = node);
	handleMouseEnter = () => {
		this.setState({
			hovered: true
		});
	};
	handleMouseLeave = () => {
		this.setState({
			hovered: false
		});
	};
	componentDidMount() {
		this.mouseEnterListener = this.ref.addEventListener(
			"mouseenter",
			this.handleMouseEnter
		);
		this.mouseLeaveListener = this.ref.addEventListener(
			"mouseleave",
			this.handleMouseLeave
		);
	}
	componentWillUnmount() {
		this.ref.removeEventListener("mouseenter", this.mouseEnterListener);
		this.ref.removeEventListener("mouseleave", this.mouseLeaveListener);
	}
	render() {
		return (
			<div ref={this.setRef}>
				{React.cloneElement(this.props.children, this.state)}
			</div>
		);
	}
}

Hoverable.propTypes = {
	children: PropTypes.element.isRequired
};
