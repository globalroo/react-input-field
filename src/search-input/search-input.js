import * as React from "react";
import PropTypes from "prop-types";
import { assignStyles } from "src/utils/assign-styles";
import {
	BORDER_WIDTH,
	BUTTON_WIDTH,
	INPUT_LEFT_OFFSET,
	styles
} from "src/search-input/styles/styles";

import { Hoverable } from "src/search-input/hoverable-component";
import { ClearButton } from "src/search-input/clear-button";

export class SearchInput extends React.Component {
	constructor(props) {
		super(props);
		this.inputRef = null;
		this.borderRadius = Math.floor(this.props.height / 2);
		this.inputWidth =
			this.props.width - BORDER_WIDTH * 2 - BUTTON_WIDTH - INPUT_LEFT_OFFSET;
		this.state = {
			content: "",
			showButton: false
		};
	}

	setInputRef = node => (this.inputRef = node);
	setFocus = () => this.inputRef.focus();

	clearText = () => {
		const content = "";
		this.inputRef.value = content;
		this.setState({ content, showButton: false });
	};

	onChangeText = event => {
		const content = event.target.value;
		const showButton = content && content.length > 0 ? true : false;
		this.setState({ content, showButton });
	};

	componentDidMount() {
		this.props.whenAvailable(this.inputRef);
	}

	render() {
		return (
			<div
				onClick={this.setFocus}
				style={assignStyles([
					{
						width: this.props.width,
						height: this.props.height,
						borderRadius: this.borderRadius
					},
					styles.componentContainer
				])}
			>
				<input
					type="text"
					style={assignStyles([
						styles.inputStyles,
						{
							marginLeft: INPUT_LEFT_OFFSET,
							width: this.inputWidth
						}
					])}
					placeholder={this.props.placeholder}
					onChange={this.onChangeText}
					ref={node => (this.inputRef = node)}
				/>
				<Hoverable>
					<ClearButton
						visible={this.state.showButton}
						onClick={this.clearText} /*hovered injected from Hoverable*/
					/>
				</Hoverable>
			</div>
		);
	}
}

SearchInput.propTypes = {
	width: PropTypes.number,
	height: PropTypes.number,
	placeholder: PropTypes.string,
	whenAvailable: PropTypes.func
};
