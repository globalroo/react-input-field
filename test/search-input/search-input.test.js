import React from "react";
import * as renderer from "react-test-renderer";
import { SearchInput } from "src/search-input/search-input";

jest.mock("src/search-input/clear-button");
jest.mock("src/search-input/hoverable-component");

describe("Search Input tests", () => {
	test("Renders the component as expected", () => {
		const spy = jest.fn();
		const component = renderer.create(
			<SearchInput
				width={200}
				height={50}
				placeholder={"test"}
				whenAvailable={spy}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	test("setInputRef correctly assigns ref", () => {
		const component = new SearchInput({
			width: 200,
			height: 50,
			placeholder: "test",
			whenAvailable: jest.fn()
		});
		const inputRefMock = "test";
		component.setInputRef(inputRefMock);
		expect(inputRefMock).toBe("test");
	});

	test("setInputRef correctly assigns ref", () => {
		const component = new SearchInput({
			width: 200,
			height: 50,
			placeholder: "test",
			whenAvailable: jest.fn()
		});
		const spy = {
			focus: jest.fn()
		};
		const inputRefMock = spy;
		component.setInputRef(inputRefMock);
		component.setFocus();
		expect(spy.focus).toHaveBeenCalledTimes(1);
	});

	test("clearText blanks the input value", () => {
		const component = new SearchInput({
			width: 200,
			height: 50,
			placeholder: "test",
			whenAvailable: jest.fn()
		});
		const mockNode = {
			value: "TEST"
		};
		component.setState = jest.fn();
		component.setInputRef(mockNode);
		expect(mockNode.value).toBe("TEST");
		component.clearText();
		expect(mockNode.value).toBe("");
		expect(component.setState).toHaveBeenCalled();
	});

	test("onChangeText updates State with 0 length string - hides button", () => {
		const component = new SearchInput({
			width: 200,
			height: 50,
			placeholder: "test",
			whenAvailable: jest.fn()
		});
		const mockEvent = {
			target: {
				value: ""
			}
		};
		component.setState = jest.fn();
		component.onChangeText(mockEvent);
		expect(component.setState).toHaveBeenCalledWith({
			content: "",
			showButton: false
		});
	});

	test("onChangeText updates State with 0 length string - hides button", () => {
		const component = new SearchInput({
			width: 200,
			height: 50,
			placeholder: "test",
			whenAvailable: jest.fn()
		});
		const mockEvent = {
			target: {
				value: "TEST"
			}
		};
		component.setState = jest.fn();
		component.onChangeText(mockEvent);
		expect(component.setState).toHaveBeenCalledWith({
			content: "TEST",
			showButton: true
		});
	});
});

// onChangeText = event => {
//   const content = event.target.value;
//   const showButton = content && content.length > 0 ? true : false;
//   this.setState({ content, showButton });
// };
