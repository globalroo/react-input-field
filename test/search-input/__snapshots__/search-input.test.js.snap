// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Input tests Renders the component as expected 1`] = `
<div
  onClick={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#222222",
      "border": "2px solid #61dafb",
      "borderRadius": 25,
      "display": "flex",
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "outline": "none",
      "padding": 5,
      "width": 200,
    }
  }
>
  <input
    onChange={[Function]}
    placeholder="test"
    style={
      Object {
        "backgroundColor": "#222222",
        "border": "0px",
        "color": "white",
        "fontSize": "18px",
        "fontWeight": "bold",
        "lineHeight": "22px",
        "marginLeft": 15,
        "outline": "none",
        "width": 136,
      }
    }
    type="text"
    value=""
  />
</div>
`;
